### DISCUSSION TITLE
Risk profile? (aka, *how* is pypraws dangerous?)

### DISCUSSION THREAD

**Discussion by blaisep** [pypyr/pypyr#281] [Q&A]
I noticed the comment:
> [You're probably running this on a dev box. Seriously, don't run this anywhere close to production, or anything you actually care about.](https://github.com/pypyr/pypyr-example#22aws)
and I wonder if that is because the they are examples, or because of something more fundamental.

Describing risks:
I also wonder if there is some benefit to include a few fields like:
    - maturity (0- no measurement, 1-history, 2- used as part of a larger construct)
    - confidence (eg. the ratio of bugs to releases, or maybe the degree of reproducibility/flakiness)
    - risk vectors (eg. non-idempotent, mutable, visible secrets)

similar to how we have a test coverage metric.... or how we think about computational complexity?

**Comment by yaythomas** [comment]
hehe, so that comment is in the `pypyr-example` repo. . . these pipelines are examples that just do arbitrary things to illustrate a point or show how to do something in a practical sense. 

When I warn operators to be cautious, it's NOT because pypyr or any of the pipelines or pypyraws is doing anything malicious/bad/dangerous/flaky.

It's because, like you say @blaisep, you should be careful with running example code in a "serious" environment.

For example, here is an example that interacts with aws: https://github.com/pypyr/pypyr-example/blob/main/pipelines/aws-s3.yaml

The pipeline shows how to upload, download files with/without encryption. Now, if you already have a file `{mybucketname}/arb.txt` in s3 - you would be overwriting it with the example file. I _think_ it's hopefully pretty unlikely that the operator chooses a bucket name as input argument that clashes, and hopefully even if that happens `arb.txt` is hopefully a pretty unlikely filename. . ..

But still, as a responsible maintainer I can't be glib and hand-wave away that you _could_ theoretically end up overwriting something you didn't mean to. Thus, warnings all over the place not to run example code that could be overwriting/updating settings/files in production. :-) 

Similarly, this is why the warnings to remember to delete/clean-up test artefacts so that operators don't get surprise bills. It's not that pypyr is doing anything wrong, it's more to be sure a pipeline author doesn't forget.

The risk matrix is an interesting idea! In this case I maybe _over_-described the risk, because I really don't want, however obliquely, to contribute to someone unthinkingly doing something that inadvertently overwrites something they didn't mean to.

**Comment by yaythomas** [comment]
To answer the question in the title of your post: it's dangerous only in the same sense that any arbitrary python is dangerous - pypyr is a power-tool, like any scripting language. You can of course code it to do whatever you want - and this might well include doing dangerous and controversial things like `$ rm -rf`. By the same token that you shouldn't be running arbitrary, untested .py files in production, you also shouldn't be running arbitrary, untested pipelines in production.
